libr/search/search.c

```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+				}
 			}
 		}
-		ch = buf[pos + kw_len];
-		pos += bad_char_shift[kw->icase?tolower(ch):ch];
+		ch = buf[i + m];
+		i += bad_char_shift[kw->icase ? tolower (ch) : ch];
 	}
 
 	return false;
 }
 #endif
 
-static int maskcmp(RSearchKeyword *kw, const ut8* buf, int kwidx) {
-	int k;
-	ut8 a, b;
-
-	kwidx = kw->idx[kwidx];
-	k = (kwidx % kw->binmask_length);
-	a = *buf;
-	b = kw->bin_keyword[kwidx];
-	if (kw->icase) {
-		a = tolower (a);
-		b = tolower (b);
+static inline bool brute_force_match(RSearch *s, RSearchKeyword *kw, const ut8 *buf, int i) {
+	int j = 0;
+	if (s->distance) { // slow path, more work in the loop
+		int dist = 0;
+		if (kw->binmask_length > 0) {
+			for (; j < kw->keyword_length; j++) {
+				int k = j % kw->binmask_length;
+				ut8 a = buf[i + j], b = kw->bin_keyword[j];
+				if (kw->icase) {
+					a = tolower (a);
+					b = tolower (b);
+				}
+				if ((a & kw->bin_binmask[k]) != (b & kw->bin_binmask[k])) {
+					dist++;
+				}
+			}
+		} else if (kw->icase) {
+			for (; j < kw->keyword_length; j++) {
+				if (tolower (buf[i + j]) != tolower (kw->bin_keyword[j])) {
+					dist++;
+				}
+			}
+		} else {
+			for (; j < kw->keyword_length; j++) {
+				if (buf[i + j] != kw->bin_keyword[j]) {
+					dist++;
+				}
+			}
+		}
+		return dist <= s->distance;
+	}
+
+	if (kw->binmask_length > 0) {
+		for (; j < kw->keyword_length; j++) {
+			int k = j % kw->binmask_length;
+			ut8 a = buf[i + j], b = kw->bin_keyword[j];
+			if (kw->icase) {
+				a = tolower (a);
+				b = tolower (b);
+			}
+			if ((a & kw->bin_binmask[k]) != (b & kw->bin_binmask[k])) break;
+		}
+	} else if (kw->icase) {
+		while (j < kw->keyword_length &&
+					 tolower (buf[i + j]) == tolower (kw->bin_keyword[j])) {
+			j++;
+		}
+	} else {
+		while (j < kw->keyword_length && buf[i + j] == kw->bin_keyword[j]) {
+			j++;
+		}
 	}
-	a &= kw->bin_binmask[k];
-	b &= kw->bin_binmask[k];
-	return (a == b)? 1: 0;
+	return j == kw->keyword_length;
 }
 
+// Supported search variants: binmask, icase, inverse, overlap
 R_API int r_search_mybinparse_update(void *_s, ut64 from, const ut8 *buf, int len) {
 	RSearch *s = (RSearch*)_s;
 	RSearchKeyword *kw;
 	RListIter *iter;
-	int count = 0;
+	RSearchLeftover *left;
+	int count = 0, longest = 0, i;
 
-#if USE_BMH
-	ut64 offset;
-	ut64 match_pos;
-	RSearchKeyword *kw;
 	r_list_foreach (s->kws, iter, kw) {
-		offset = 0;
-		while (offset < len && r_search_bmh (kw, offset, buf, len, &match_pos)) {
-			if (!r_search_hit_new (s, kw, from + match_pos)) {
-				eprintf ("Something very bad has happened...\n");
-				return -1;
-			}
-			offset += match_pos + kw->keyword_length;
-			kw->count++;
-			count++;
-			/* Stop at the first occurrence */
-			if (s->inverse) {
-				return -1;
-			}
+		longest = R_MAX (longest, kw->keyword_length);
+	}
+	if (!longest) {
+		return 0;
+	}
+	if (s->data) {
+		left = s->data;
+		if (left->end != from) {
+			left->len = 0;
+		}
+	} else {
+		left = malloc (sizeof(RSearchLeftover) + 2 * (longest - 1));
```

Comment:
Changed to `		left = malloc (sizeof(RSearchLeftover) + (size_t)2 * (longest - 1));`

It will never cause trouble where `size_t` is 64-bit and `longest` is 32-bit.